### ISSUE TITLE
Assert doesn't erase nilable when used in branch tail...

### ISSUE THREAD

**Issue by jcorbin** [aardappel/lobster#79]
Rather it only seems to work when it's in the main branch.

This fails:
```
class such:
    x = 0

var much = nil

if not much:
    much = such {}
    assert much

much.x++
print(much)

much.x++
print(much)
```

With `foo.lobster(10): error: type such? does not have field: x`

But hoisting the `assert much` works.

This has been a contrived  example, reduced down from something like:
- a class with a nil-able `resource?` field
- where I load it the first time through, something like `if not tex: load()`
- where there's a `def load()` method that ends with an `assert tex` after having loaded from a file
- I ended up just inlining the loader method, and hoisting its guard assertion into main control flow...

**Comment by aardappel** [comment]
Yeah, there's currently no flow information going *out* of scopes. If there was, the `assert` wouldn't even be necessary, since `much = such {}` already makes it non-nil.

But yeah, this would require interesting logic, since now we need to basically say, both "branches" of the `if` come to the same conclusion for a certain variable, so that information can be merged. That would be a new kind of construct in the typechecker.

I guess I haven't run into this much, since the nil checking just makes me religiously try to avoid any nils in my code :)

**Comment by aardappel** [comment]
This is fixed nowadays:

```
test.lobster(8): warning: assert will always succeed with non-nil reference type `such`
such{1}
such{2}
```

i.e. it even works without the `assert`
